### DISCUSSION TITLE
Work around Cloudflare issues

### DISCUSSION THREAD

**Discussion by ghost** [FreshRSS/FreshRSS#7229] [Q&A]
**Is your feature request related to a problem? Please describe.**
An increasingly large number of feeds are blocked by Cloudflare due to misconfiguations by web developers. Unfortunately, it's becoming more and more common and it's making great tools like FreshRSS less useful, at no fault of the software.  

**Describe the solution you’d like**
I don't know exactly what it is I'm suggesting or how this should be approached, if it even can be.  So consider this more of a potential feature discussion rather than a request.  There IS a solution to this problem, but it comes in the form of a Python library.  I wanted to open up a discussion on the topic to see if there's anything FreshRSS could do with this.

There is a python library called [cloudscraper](https://github.com/VeNoMouS/cloudscraper) that works brilliantly for getting around Cloudflare restrictions.  There was a [node js library](https://github.com/codemanki/cloudscraper) of the same name, but it is now defunct.

From the cloudscraper readme:
> Cloudflare's anti-bot page currently just checks if the client supports Javascript, though they may add additional techniques in the future.
>
> Due to Cloudflare continually changing and hardening their protection page, cloudscraper requires a JavaScript Engine/interpreter to solve Javascript challenges. This allows the script to easily impersonate a regular web browser without explicitly deobfuscating and parsing Cloudflare's Javascript.

Are there any options available to implement something like this.. either natively or via extension?

**Describe alternatives you’ve considered**
.

**Additional context**
.

**Comment by squromiv** [comment]
#1646

**Comment by squromiv** [comment]
> Are there any options available to implement something like this.. either natively or via extension?

Any news? I have spent a lot of time trying to find a solution.

**Comment by squromiv** [comment]
There is quite similar [discussion](https://github.com/RSS-Bridge/rss-bridge/issues/2547) in RSSBridge repository. Let's join forces! @aledeg @math-GH @0x199x Please, take a look at [curl-impersonate](https://github.com/lwthiker/curl-impersonate) project. Any thoughts?

**Comment by Alkarex** [comment]
Another example of problem https://github.com/FreshRSS/FreshRSS/discussions/4556

**Comment by squromiv** [comment]
@Alkarex 
So, is there any chance to include this solution to FreshRSS?

**Comment by Alkarex** [comment]
> @Alkarex So, is there any chance to include this solution to FreshRSS?

If anyone is motivated to develop an extension, that would be welcomed. I can support with additional extension hooks, should that be necessary.

**Comment by ryancom16** [comment]
Hi! I have a working solution that I  have been using for a couple years and could be implemented into an extension. I use 
[FlareSolverr](https://github.com/FlareSolverr/FlareSolverr). FlareSolver is able to render pages protected by cloudflare. 

I run both FreshRSS and FlareSolver in separate docker containers. 

The biggest barrier between FlareSolverr and FreshRSS is that FlareSolverr needs a post request with some specific details. The response is also needs a bit of parsing. 

To bridge this capability gap, I wrote a little php code that that takes a feed url as a parameter and then makes a proper post request to FlareSolverr and then returns the HTML response. 

Here is the workflow:
1. FreshRss  is configured with a new feed pointing to SERVERIP/FlareSolverr/FlareSolverrRequest.php?feed=CF PROTECTED FEED
2. FlareSolverrRequest.php does a post request to the flaresolverr docker using the following:

```php
<?php
//Returns request for website protected by Cloudflare
//Requires  ghcr.io/flaresolverr/flaresolverr docker container
$feed = $_GET['feed'];
$session = $_GET['session'];
$ch = curl_init();
$headers  = [
    'Content-Type: application/json'
];
$postData = [
    'cmd' => 'request.get',
    'url' => $feed,
    'session' => $session
];
//
curl_setopt($ch, CURLOPT_URL, "http://localhost:8191/v1"); //This is my flaresolverr address 
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36');
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($postData));
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

$array = json_decode(curl_exec($ch), true);

echo $array['solution']['response'];

curl_close($ch); 
```

3. Flaresolver returns HTML/xml response to FlareSolverrRequest.php which is then sent back to FreshRSS. 

I have had incredible consistency by using this.

Here is a description of the extension:
1. Extension configuration will allow user to configure the address for the locally or publicly hosted FlareSolverr address. 
2. Every feed will have the option to enable the Cloudflare bypass. 
3. If enabled, the extension code will perform the post request to the specified address and then return the response. 

Hope this helps :)

**Comment by ryancom16** [comment]
Check out [Prowlarr](https://github.com/Prowlarr/Prowlarr) and how they implement the cloudflare bypass.

**Comment by Alkarex** [comment]
Looks good @ryancom16 👍🏻

**Comment by SMylk** [comment]
Is there a way to add this to the docker image?

**Comment by Alkarex** [comment]
@SMylk Most likely as a FreshRSS extension, and with flaresolverr added to docker-compose

**Comment by SMylk** [comment]
I reckon there is no extension yet then :(

**Comment by Alkarex** [comment]
Contributions welcome ;-)

**Comment by SMylk** [comment]
Solved my issue by sending feeds through feedburner, not ideal, but it works.

**Comment by SMylk** [comment]
> Hi! I have a working solution that I have been using for a couple years and could be implemented into an extension. I use [FlareSolverr](https://github.com/FlareSolverr/FlareSolverr). FlareSolver is able to render pages protected by cloudflare.
> 
> I run both FreshRSS and FlareSolver in separate docker containers.
> 
> The biggest barrier between FlareSolverr and FreshRSS is that FlareSolverr needs a post request with some specific details. The response is also needs a bit of parsing.
> 
> To bridge this capability gap, I wrote a little php code that that takes a feed url as a parameter and then makes a proper post request to FlareSolverr and then returns the HTML response.
> 
> Here is the workflow:
> 
>     1. FreshRss  is configured with a new feed pointing to SERVERIP/FlareSolverr/FlareSolverrRequest.php?feed=CF PROTECTED FEED
> 
>     2. FlareSolverrRequest.php does a post request to the flaresolverr docker using the following:
> 
> 
> ```
> <?php
> //Returns request for website protected by Cloudflare
> //Requires  ghcr.io/flaresolverr/flaresolverr docker container
> $feed = $_GET['feed'];
> $session = $_GET['session'];
> $ch = curl_init();
> $headers  = [
>     'Content-Type: application/json'
> ];
> $postData = [
>     'cmd' => 'request.get',
>     'url' => $feed,
>     'session' => $session
> ];
> //
> curl_setopt($ch, CURLOPT_URL, "http://localhost:8191/v1"); //This is my flaresolverr address 
> curl_setopt($ch, CURLOPT_POST, 1);
> curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
> curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
> curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36');
> curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($postData));
> curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
> 
> $array = json_decode(curl_exec($ch), true);
> 
> echo $array['solution']['response'];
> 
> curl_close($ch); 
> ```
> 
>     3. Flaresolver returns HTML/xml response to FlareSolverrRequest.php which is then sent back to FreshRSS.
> 
> 
> I have had incredible consistency by using this.
> 
> Here is a description of the extension:
> 
>     1. Extension configuration will allow user to configure the address for the locally or publicly hosted FlareSolverr address.
> 
>     2. Every feed will have the option to enable the Cloudflare bypass.
> 
>     3. If enabled, the extension code will perform the post request to the specified address and then return the response.
> 
> 
> Hope this helps :)

This is good, but i get "This does not appear to be a valid RSS or Atom feed. " for every feed I try, however it can load it and I see the content in the browser.

**Comment by ravenscroftj** [comment]
Hi all,

I came across this thread and I was inspired by @ryancom16's solution to write a full plugin for FreshRSS which you can find here:
https://github.com/ravenscroftj/freshrss-flaresolverr-extension

I initially had the same problem as @SMylk - it turns out that the headless browser inside Flaresolverr injects style information to RSS feeds so I added some logic that strips this away and just produces a pure RSS feed using PHP's `DOMDocument`.

**Comment by enchained** [comment]
@ravenscroftj Is there a way to make it work with the website url used as a feed url? I use web scraping as a type of feed source in FreshRSS, and passing the url through this plugin causes `error on line 2 at column 1: Start tag expected, '<' not found` error.

**Comment by ryancom16** [comment]
@enchained I think that will only work if you enter the url of a website that returns a feed, usually ending in .rss. The plugin is looking for a specific format. 

So, I think what you are looking to do is create a custom RSS feed out of website that is protected by CF. For this, you can use [RSS-Bridge](https://github.com/RSS-Bridge/rss-bridge) and create your own bridge. Refer to the documentation, but this has worked extremely well for me and works extremely well with FreshRSS. What you would do is create a custom bridge for your website. Each "bridge" is coded to retrieve the web information, parse it, and then format it as an RSS feed. I have a FlareSolverr instance running in a docker. When my custom bridge queries a CF protected site, I send that request to my FlareSolverr and it returns the HTML that I can parse. 

Note that my above code that inspired the extension is now out of date. FlareSolverr has been updated since. 

In the bridge I request the website contents in this way
```
public $uri = "http://<Your Web Server>/FlareSolverr/FlareSolverrv3.php?session=loadbridge&session_timeout=60&url=<Website URL protected by CF>";

$html = getSimpleHTMLDOM($urli);
```

FlareSolverr3.php

```
<?php
//Returns request for website protected by Cloudflare
//Requires  ghcr.io/flaresolverr/flaresolverr docker container
$url = $_GET['url'];
$session = $_GET['session'];
$sessiontimeout = $_GET['sessiontimeout'];
$ch = curl_init();
$headers  = [
    'Content-Type: application/json'
];
$postData = [
    'cmd' => 'request.get',
    'url' => $url,
    'session' => $session,
    'session_ttl_minutes' => 60,
    'maxTimeout' => 60000
];


//'userAgent' => 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36',
//'maxTimeout' => 60000
curl_setopt($ch, CURLOPT_URL, "http://localhost:8191/v1");
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/62.0.3202.94 Safari/537.36');
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($postData));
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);

$array = json_decode(curl_exec($ch), true);
//echo curl_exec($ch);
//echo $array['solution'];
echo $array['solution']['response'];

//$html = curl_exec($ch);
//file_put_contents(__DIR__ . '/output.html', $array['solution']['response']);


curl_close($ch);
```

However, I believe I could just call FlareSolverr directly from the bridge itself if I wanted to.

After you have your working feed in you copy of RSS-Bridge, you can added it to FreshRSS

**Reply by juju4** [reply]
An updated version with logging, proxy and error handling
```
<?php
// Adapted from https://github.com/FreshRSS/FreshRSS/issues/4323#issuecomment-1304715559
// Returns request for website protected by Cloudflare
// Requires  ghcr.io/flaresolverr/flaresolverr docker container
$debug = false;
$proxy_enabled = false;
$proxy_url = "http://10.1.2.3:3128";
$proxy_errors = [ 'The requested URL could not be retrieved', 'The following error was encountered while trying to retrieve the URL:'];

if (PHP_SAPI === 'cli') {
    $url = strval($argv[1]);
    $ip = '-';
    $ua = 'cli';
    define('LOG_DIR', '/tmp');
} else {
    $url = strval($_GET['url']);
    $ip = $_SERVER['REMOTE_ADDR'];
    $ua = $_SERVER['HTTP_USER_AGENT'];
    define('LOG_DIR', '/var/www/log');
}

if (!parse_url($url) && !filter_var($url, FILTER_VALIDATE_URL) && !str_starts_with($url, 'http')) {
    header($_SERVER["SERVER_PROTOCOL"].' 404 Invalid feed url ');
    die();
}
define('LOG_FILE', LOG_DIR.'/flaresolverr.log');
$today = date("Y-m-d\TH:i:sP");
file_put_contents(LOG_FILE, "time=$today msg=request ip=$ip url=$url ua=$ua\n", FILE_APPEND | LOCK_EX);

$session = '';
if (!empty($_GET['session'])) {
  $session = strval($_GET['session']);
}
$sessiontimeout = 60000;
if (!empty($_GET['sessiontimeout'])) {
  $sessiontimeout = intval($_GET['sessiontimeout']);
}
$ch = curl_init();
$headers  = [
    'Content-Type: application/json'
];
if ($proxy_enabled === true) {
  $postData = [
    'cmd' => 'request.get',
    'url' => $url,
    'session' => $session,
    'session_ttl_minutes' => 60,
    'maxTimeout' => $sessiontimeout,
    'proxy' => [ 'url' => $proxy_url ]
  ];
} else {
  $postData = [
    'cmd' => 'request.get',
    'url' => $url,
    'session' => $session,
    'session_ttl_minutes' => 60,
    'maxTimeout' => $sessiontimeout
  ];
}

curl_setopt($ch, CURLOPT_URL, "http://localhost:8191/v1"); // Flaresolverr address
curl_setopt($ch, CURLOPT_POST, 1);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36');
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($postData));
curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 0);
curl_setopt($ch, CURLOPT_TIMEOUT, 300); //timeout in seconds

$curl_ret = curl_exec($ch);
if( ! $curl_ret = curl_exec($ch)) {
  file_put_contents(LOG_FILE, "time=$today msg=error ip=$ip url=$url ua=$ua error=".strval(curl_error($ch))."\n", FILE_APPEND | LOCK_EX);
} else {
  $array = json_decode($curl_ret, true);
  if (!empty($array['solution']['response']) and $array['solution']['response'] != '<html><head></head><body></body></html>') {
    $log_suffix = '';
    // catch-up proxy errors as return code returns 200...
    if ($proxy_enabled === true and preg_match('/'.join('|',$proxy_errors).'/',$array['solution']['response'])) {
        $log_suffix = ' proxy_enabled=$proxy_enabled proxy_url=$proxy_url proxymsg=error';
    } elseif {
        $log_suffix = ' proxy_enabled=$proxy_enabled proxy_url=$proxy_url';
    }

    echo $array['solution']['response'];

    if ($debug) {
        file_put_contents(LOG_DIR . '/output.html', $array['solution']['response']);
        file_put_contents(LOG_DIR. '/output.json', $curl_ret);
    }
    // FIXME! flaresolverr returns empty headers field. Selenium issue
    // https://github.com/FlareSolverr/FlareSolverr/issues/1162
    // https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/141
    // and no response size
    $today = date("Y-m-d\TH:i:sP");
    file_put_contents(LOG_FILE, "time=$today msg=completed ip=$ip url=$url ua=$ua status_code=".intval($array['solution']['status'])." flaresolverr_status=".$array['status']." flaresolverr_msg=".$array['message']." flaresolverr_version=".$array['version']."$log_suffix\n", FILE_APPEND | LOCK_EX);
  } else {
    file_put_contents(LOG_FILE, "time=$today msg=empty response ip=$ip url=$url ua=$ua status_code=".intval($array['solution']['status'])." flaresolverr_status=".$array['status']." flaresolverr_msg=".$array['message']." flaresolverr_version=".$array['version']."$log_suffix\n", FILE_APPEND | LOCK_EX);
    header($protocol.' 204 No Content/Empty response');
    echo "Empty response from $url";
    die();
  }
}

curl_close($ch);

?>
```

**Comment by Alkarex** [comment]
Related guide for server admins https://openrss.org/blog/using-cloudflare-on-your-website-could-be-blocking-rss-users
We should recommend a guide that could be sent to the Web master of a problematic feed. Input welcome

**Reply by Alkarex** [reply]
And in French: https://www.abyssproject.net/2025/12/autoriser-les-lecteurs-rss-derriere-cloudflare/

**Comment by rwiesler** [comment]
It might be an idea to add FreshRSS as an verfied Bot in the [Cloudflare Bots Directory](https://radar.cloudflare.com/bots/directory?q=rss&kind=all)? (here paired with the search term 'rss')

> [...] Web Bot Auth is an authentication method that leverages cryptographic signatures in HTTP messages to verify that a request comes from an automated bot. Web Bot Auth is used as a verification method for [verified bots](https://developers.cloudflare.com/bots/concepts/bot/verified-bots/) and [signed agents](https://developers.cloudflare.com/bots/concepts/bot/signed-agents/).
>
>It relies on two active IETF drafts: a [directory draft ↗](https://datatracker.ietf.org/doc/html/draft-meunier-http-message-signatures-directory) allowing the crawler to share their public keys, and a [protocol draft ↗](https://datatracker.ietf.org/doc/html/draft-meunier-web-bot-auth-architecture) defining how these keys should be used to attach crawler's identity to HTTP requests. 
>
>This documentation goes over specific integration within Cloudflare. [...]

See further [Cloudflares reference documentation regarding Web Bot Auth](https://developers.cloudflare.com/bots/reference/bot-verification/web-bot-auth/).

**Reply by Alkarex** [reply]
I do not think this is applicable to decentralised, self-hosted applications such as FreshRSS, but if so, more information welcome
